src/datadog/remote_config/remote_config.cpp

```diff
@@ -230,41 +240,47 @@ void Manager::process_response(const nlohmann::json& json) {
         return;
       }
 
-      auto decoded_config =
-          base64_decode(target_it.value().at("raw").get<StringView>());
-
-      const auto config_json = nlohmann::json::parse(decoded_config);
+      auto raw_data = target_it->at("raw").get<StringView>();
+      auto decoded_config = base64_decode(raw_data);
 
       Configuration new_config;
-      new_config.id = config_json.at("id");
+      new_config.id = config_key_data->config_id;
       new_config.path = std::string{config_path};
       new_config.hash = config_metadata.at("/hashes/sha256"_json_pointer);
       new_config.content = std::move(decoded_config);
-      new_config.version = config_json.at("revision");
-      new_config.product = *product;
-
-      const auto& targeted_service = config_json.at("service_target");
-      if (targeted_service.at("service").get<StringView>() !=
-              tracer_signature_.default_service ||
-          targeted_service.at("env").get<StringView>() !=
-              tracer_signature_.default_environment) {
-        new_config.state = Configuration::State::error;
-        new_config.error_message = "Wrong service targeted";
-      } else {
-        for (const auto& listener : listeners_per_product_[*product]) {
-          // Q: Two listeners on the same product. What should be the behaviour
-          // if one of the listeners report an error?
-          // R(@dmehala): Unspecified. For now, the config is marked with an
-          // error.
-          if (auto error_msg = listener->on_update(new_config)) {
-            new_config.state = Configuration::State::error;
-            new_config.error_message = std::move(*error_msg);
-          } else {
-            new_config.state = Configuration::State::acknowledged;
-          }
+      new_config.version = config_metadata.at("/custom/v"_json_pointer);
+      new_config.product = product;
+
+      // TODO: should be moved to the listener
+      if (product == product::Flag::APM_TRACING) {
+        const auto config_json = nlohmann::json::parse(new_config.content);
+        new_config.version = config_json.at("revision");
+
+        const auto& targeted_service = config_json.at("service_target");
+        if (targeted_service.at("service").get<StringView>() !=
+                tracer_signature_.default_service ||
+            targeted_service.at("env").get<StringView>() !=
+                tracer_signature_.default_environment) {
+          new_config.state = Configuration::State::error;
+          new_config.error_message = "Wrong service targeted";
+          goto skip_listeners;
+        }
```

Comment:
Does that mean outside of APM_TRACING product, other services do not target `service` and `env`?

Reply:
Remote config provides configuration only for the service/env that the remote config request specifies (client/client_tracer/{service,env})

Reply:
The remote config request always include the service and env ([code for ref](https://github.com/DataDog/dd-trace-cpp/blob/aa037a9cdea70d50a193ff42173fa84038b1c24e/src/datadog/remote_config/remote_config.cpp#L107-L113). If understand correctly, then, this change is not required. Am I missing something?

Reply:
Yes, the **request** always includes the service and env, which is how the backend decides which configuration to send for asm. But this is about processing the **response**. In particular, this is assuming that the response is a json file and that it includes `service` and `env` keys. This is only true for APM_TRACING configuration.

Reply:
Ok. Pardon me, I want to make sure I understand. Does that mean for ASM, the response is not necessarily a json file?

Reply:
It is a json file, yes. But it doesn't have any `service_target` key. Plus, duplicativeat the RC protocol level, the configuration need not be json.

The best thing to do here would be to move the block to the APM_TRACING listener. Notethat parsing the content as json at this point is duplicative; the configuration passed to the listener is in the form of std::string. So the listener needs to parse the content as json again.